File: backend/app/core/pricing.py
from calendar import monthrange
from dataclasses import dataclass
from datetime import datetime, timezone
from decimal import Decimal

from app.core.money import money


@dataclass
class MotorcycleRates:
    daily: Decimal
    three_day: Decimal
    weekly: Decimal
    monthly: Decimal


@dataclass
class LineAmounts:
    charge: Decimal
    discount: Decimal
    line_total: Decimal


@dataclass
class DocumentTotals:
    subtotal: Decimal
    discount: Decimal
    tax_percent: Decimal
    tax: Decimal
    total: Decimal


def resolve_motorcycle_rates(
    daily_rate=None, three_day_rate=None, weekly_rate=None, monthly_rate=None
) -> MotorcycleRates:
    daily = money(daily_rate) if daily_rate and daily_rate > 0 else Decimal("0.00")

    def positive(value) -> Decimal:
        try:
            v = money(value)
            return v if v > 0 else Decimal("0.00")
        except Exception:
            return Decimal("0.00")

    three_day = positive(three_day_rate) or money(daily * 3)
    weekly = positive(weekly_rate) or money(daily * Decimal("6.5"))
    monthly = positive(monthly_rate) or money(daily * 22)
    return MotorcycleRates(daily=daily, three_day=three_day, weekly=weekly, monthly=monthly)


def duration_days(start: datetime, due: datetime) -> int:
    if not start or not due:
        return 0
    if start.tzinfo is None:
        start = start.replace(tzinfo=timezone.utc)
    if due.tzinfo is None:
        due = due.replace(tzinfo=timezone.utc)
    seconds = (due - start).total_seconds()
    if seconds <= 0:
        return 0
    return max(int(seconds / 86400 + 0.999999), 1) if seconds % 86400 else int(seconds // 86400)


def _at_minute(value: datetime) -> datetime:
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.replace(second=0, microsecond=0)


def add_months(start: datetime, months: int) -> datetime:
    """Add calendar months, clamping to the last day of the target month."""
    month_index = start.month - 1 + max(0, int(months or 0))
    year = start.year + month_index // 12
    month = month_index % 12 + 1
    day = min(start.day, monthrange(year, month)[1])
    return start.replace(year=year, month=month, day=day)


def calendar_months_between(start: datetime | None, due: datetime | None) -> int:
    """Return n when due is exactly start plus n calendar months (minute precision)."""
    if not start or not due:
        return 0
    start_m = _at_minute(start)
    due_m = _at_minute(due)
    if due_m <= start_m:
        return 0
    for n in range(1, 37):
        if _at_minute(add_months(start_m, n)) == due_m:
            return n
    return 0


def rate_type_for(days: int, start: datetime | None = None, due: datetime | None = None) -> str:
    months = calendar_months_between(start, due)
    if months >= 1:
        return "Monthly"
    d = max(0, int(days or 0))
    if d == 3:
        return "ThreeDay"
    if d == 7:
        return "Weekly"
    if 28 <= d <= 31:
        return "Monthly"
    return "Daily"


def line_charge(
    rates: MotorcycleRates,
    days: int,
    start: datetime | None = None,
    due: datetime | None = None,
) -> Decimal:
    d = max(0, int(days or 0))
    if d <= 0:
        return Decimal("0.00")
    months = calendar_months_between(start, due)
    if months >= 1:
        return money(rates.monthly * months)
    if d == 1:
        return rates.daily
    if d == 3:
        return rates.three_day
    if d == 7:
        return rates.weekly
    if 28 <= d <= 31:
        return rates.monthly
    return money(rates.daily * d)


def line_amounts(
    rates: MotorcycleRates, days: int, discount=0, start: datetime | None = None, due: datetime | None = None
) -> LineAmounts:
    charge = line_charge(rates, days, start, due)
    disc = min(max(money(discount), Decimal("0.00")), charge)
    return LineAmounts(charge=charge, discount=money(disc), line_total=money(max(charge - disc, Decimal("0"))))


def document_totals(line_totals: list[Decimal], discount=0, tax_percent=0) -> DocumentTotals:
    subtotal = money(sum(line_totals, Decimal("0")))
    disc = min(max(money(discount), Decimal("0.00")), subtotal)
    taxable = max(subtotal - disc, Decimal("0"))
    pct = max(money(tax_percent), Decimal("0"))
    tax = money(taxable * pct / Decimal("100"))
    return DocumentTotals(subtotal=subtotal, discount=money(disc), tax_percent=pct, tax=tax, total=money(taxable + tax))


def suggested_deposit(rates: MotorcycleRates) -> Decimal:
    return money(max(rates.daily * 10, 50))

File: frontend/app/utils/rental/pricing.ts
/** Motorcycle rate-tier pricing for rental create / return totals. */

export interface MotorcycleRates {
  dailyRate?: unknown
  threeDayRate?: unknown
  weeklyRate?: unknown
  monthlyRate?: unknown
  [key: string]: unknown
}

export interface RentalLineAmounts {
  charge: number
  discount: number
  lineTotal: number
}

export interface DocumentTotals {
  subtotal: number
  discount: number
  taxPercent: number
  tax: number
  total: number
}

export type RentalRatePlan = '1d' | '3d' | '1w' | '1m' | 'custom'
export type RentalRateType = 'Daily' | 'ThreeDay' | 'Weekly' | 'Monthly'

function asMoney(value: unknown) {
  const n = Number(value)
  return Number.isFinite(n) && n > 0 ? n : 0
}

function round2(value: number) {
  return Number(value.toFixed(2))
}

function pad2(n: number) {
  return String(n).padStart(2, '0')
}

function parseDateTime(value: string) {
  if (!value) return null
  const date = new Date(value)
  return Number.isFinite(date.getTime()) ? date : null
}

export function formatDateTimeLocal(date: Date): string {
  return `${date.getFullYear()}-${pad2(date.getMonth() + 1)}-${pad2(date.getDate())}T${pad2(date.getHours())}:${pad2(date.getMinutes())}`
}

function dateTimeMinuteKey(value: string) {
  const date = parseDateTime(value)
  return date ? formatDateTimeLocal(date) : ''
}

/** Resolve tier amounts with sensible fallbacks from dailyRate. */
export function resolveMotorcycleRates(moto: MotorcycleRates | null | undefined) {
  const daily = asMoney(moto?.dailyRate)
  const threeDay = asMoney(moto?.threeDayRate) || round2(daily * 3)
  const weekly = asMoney(moto?.weeklyRate) || round2(daily * 6.5)
  const monthly = asMoney(moto?.monthlyRate) || round2(daily * 22)
  return { daily, threeDay, weekly, monthly }
}

/** Add calendar months, clamping to the last day of the target month (Jan 31 → Feb 28/29). */
export function addMonthsToDateTime(start: string, months: number): string {
  const base = parseDateTime(start)
  if (!base) return ''
  const m = Math.max(0, Math.floor(Number(months) || 0))
  const day = base.getDate()
  const result = new Date(base)
  result.setDate(1)
  result.setMonth(result.getMonth() + m)
  const lastDay = new Date(result.getFullYear(), result.getMonth() + 1, 0).getDate()
  result.setDate(Math.min(day, lastDay))
  return formatDateTimeLocal(result)
}

/** Exact calendar-month count when due matches start plus n months at the same clock time. */
export function calendarMonthsBetween(start: string, due: string): number {
  const dueKey = dateTimeMinuteKey(due)
  if (!start || !dueKey) return 0
  for (let n = 1; n <= 36; n++) {
    if (dateTimeMinuteKey(addMonthsToDateTime(start, n)) === dueKey) return n
  }
  return 0
}

export function detectRatePlan(start: string, due: string, days = 0): RentalRatePlan {
  if (start && due && calendarMonthsBetween(start, due) === 1) return '1m'
  const d = days || (start && due ? daysBetween(start, due) : 0)
  if (d === 1) return '1d'
  if (d === 3) return '3d'
  if (d === 7) return '1w'
  return 'custom'
}

export function dueDateFromRatePlan(start: string, plan: RentalRatePlan, days = 1): string {
  if (!start) return ''
  if (plan === '1m') return addMonthsToDateTime(start, 1)
  if (plan === '1w') return addDaysToDateTime(start, 7)
  if (plan === '3d') return addDaysToDateTime(start, 3)
  if (plan === '1d') return addDaysToDateTime(start, 1)
  return addDaysToDateTime(start, Math.max(1, Math.floor(Number(days) || 1)))
}

export function rentalRateType(days: number, start = '', due = ''): RentalRateType {
  const months = start && due ? calendarMonthsBetween(start, due) : 0
  if (months >= 1) return 'Monthly'
  const d = Math.max(0, Math.floor(Number(days) || 0))
  if (d === 3) return 'ThreeDay'
  if (d === 7) return 'Weekly'
  if (d >= 28 && d <= 31) return 'Monthly'
  return 'Daily'
}

/**
 * Package rate shown in the unit-price field: daily / 3-day / weekly / monthly.
 * Custom durations use the daily rate except the 28–31 day monthly band.
 */
export function appliedUnitPrice(
  moto: MotorcycleRates | null | undefined,
  days: number,
  start = '',
  due = '',
): number {
  const rates = resolveMotorcycleRates(moto)
  const months = start && due ? calendarMonthsBetween(start, due) : 0
  if (months >= 1) return rates.monthly
  const d = Math.max(0, Math.floor(Number(days) || 0))
  if (d === 3) return rates.threeDay
  if (d === 7) return rates.weekly
  if (d >= 28 && d <= 31) return rates.monthly
  return rates.daily
}

/**
 * Charge for a duration using motorcycle packages:
 * 1 day, 3 days, 1 week, and calendar month(s) when start/due match.
 * 28–31 day spans still use monthly_rate when they are not an exact calendar month.
 * Other durations use dailyRate × days.
 */
export function lineCharge(
  moto: MotorcycleRates | null | undefined,
  days: number,
  start = '',
  due = '',
): number {
  const d = Math.max(0, Math.floor(Number(days) || 0))
  if (d <= 0) return 0
  const rates = resolveMotorcycleRates(moto)
  const months = start && due ? calendarMonthsBetween(start, due) : 0
  if (months >= 1) return round2(rates.monthly * months)
  if (d === 1) return rates.daily
  if (d === 3) return rates.threeDay
  if (d === 7) return rates.weekly
  if (d >= 28 && d <= 31) return rates.monthly
  return round2(rates.daily * d)
}

export function lineAmounts(
  moto: MotorcycleRates | null | undefined,
  days: number,
  discount: number = 0,
  start = '',
  due = '',
): RentalLineAmounts {
  const charge = lineCharge(moto, days, start, due)
  const disc = Math.min(Math.max(0, Number(discount) || 0), charge)
  return {
    charge,
    discount: round2(disc),
    lineTotal: round2(Math.max(charge - disc, 0)),
  }
}

/** Suggest deposit from motorcycle daily rate. */
export function suggestedDeposit(moto: MotorcycleRates | null | undefined) {
  const daily = resolveMotorcycleRates(moto).daily
  return Math.max(Math.round(daily * 10), 50)
}

export function documentTotals(input: {
  lineTotals: number[]
  discount?: number
  taxPercent?: number
}): DocumentTotals {
  const subtotal = round2(input.lineTotals.reduce((sum, n) => sum + (Number(n) || 0), 0))
  const discount = Math.min(Math.max(0, Number(input.discount) || 0), subtotal)
  const taxable = Math.max(subtotal - discount, 0)
  const taxPercent = Math.max(0, Number(input.taxPercent) || 0)
  const tax = round2(taxable * taxPercent / 100)
  return {
    subtotal,
    discount: round2(discount),
    taxPercent,
    tax,
    total: round2(taxable + tax),
  }
}

/** Allocate a payment amount across line totals by share (last line gets remainder). */
export function allocateRentalPayment(lineTotals: number[], paidAmount: number): number[] {
  const total = lineTotals.reduce((sum, n) => sum + n, 0)
  const paid = Math.max(0, Number(paidAmount) || 0)
  if (total <= 0 || paid <= 0 || !lineTotals.length) return lineTotals.map(() => 0)
  const capped = Math.min(paid, total)
  const shares = lineTotals.map((line, index) => {
    if (index === lineTotals.length - 1) return 0
    return round2(capped * (line / total))
  })
  const allocated = shares.reduce((sum, n) => sum + n, 0)
  shares[shares.length - 1] = round2(capped - allocated)
  return shares
}

/** Days between two datetime-local strings (ceil, min 1 when positive). */
export function daysBetween(start: string, due: string): number {
  if (!start || !due) return 0
  const ms = new Date(due).getTime() - new Date(start).getTime()
  if (!Number.isFinite(ms) || ms <= 0) return 0
  return Math.max(Math.ceil(ms / 86400000), 1)
}

/** Add days to a datetime-local string, preserving time portion. */
export function addDaysToDateTime(start: string, days: number): string {
  const base = parseDateTime(start)
  if (!base) return ''
  const d = Math.max(0, Math.floor(Number(days) || 0))
  base.setDate(base.getDate() + d)
  return formatDateTimeLocal(base)
}

export function todayDateTimeLocal(date = new Date()): string {
  return formatDateTimeLocal(date)
}
